### DISCUSSION TITLE
[question] how to pass addtional parameters to widget

### DISCUSSION THREAD

**Discussion by bforget163** [eclipse-theia/theia#8849] [Q&A] [answered]
I'm trying to implenent a `widget` and `openhandler` that treat URIs with different query part to be the same resource and then use the query part as addtion parameters to the widget. The case is that I have two links in the tree:  
```
bb:/resource1.test?scroll=cat
bb:/resource1.test?scroll=dog
```
and when the link been clicked/switched there should open the same widget and do the lazy load for cat or dot data and scroll to it.  
I tried to override the open() method from `WidgetOpenHandler` but it somehow breaks theia's `storeState` behavior.
```
    async open(uri: URI, options?: WidgetOpenerOptions): Promise<ResourceWidget> {
        const widget = await this.getOrCreateWidget(uri, options);
        await this.doOpen(widget, options);
        const { scroll } = qs.parse(uri);
        widget.scroll= uri;
        return widget;
    }
```
Is there a right way of doing this?
thanks in advance!!!

**Comment by bforget163** [comment, marked as answer]
I finally resolved this by override the `getOrCreateWidget` method from `WidgetOpenHandler`. Can anyone help confirm if this is the correct way?
```
    protected createWidgetOptions(uri: URI, options?: WidgetOpenerOptions): ResourceWidgetOptions {
        const mark = uri.withoutQuery().withoutFragment().toString();
        return { mark };
    }
    protected async getOrCreateWidget(uri: URI, options?: WidgetOpenerOptions): Promise<ResourceWidget> {
        const widget = await super.getOrCreateWidget(uri, options);
        widget.Uri = uri;
        return widget;
    }
```

**Reply by kittaakos** [reply]
Nice solution👍 I would customize the `WidgetManager`, though.

Just wanted to send you the link, that widget `options` is used to determine a unique "key" for a widget:
https://github.com/eclipse-theia/theia/blob/03a8cdb0566bc179f1a30c948ff8a104f8b89b82/packages/core/src/browser/widget-manager.ts#L198-L202
([Here](https://community.theia-ide.org/t/how-can-i-create-widgets-with-dynamic-ids/1302/4) is a related discussion about this topic, the other way around.)

So you need to make sure, the fragment is not involved in the widget "key" generation, and your custom widget can still dispatch some specific logic based on the URI fragment.
